active/0051-packaging.md

```diff
@@ -0,0 +1,338 @@
++ Start Date: 2017-09-07
++ RFC PR: (leave this empty)
+
+# Summary
+
+The goal of this RFC is to solidify the next version of the mechanism that
+`ember-cli` uses to build final assets. It will allow for a more flexible build
+pipeline for Ember applications. It also unlocks building experimental features
+on top. It is a backward compatible change.
+
+# Motivation
+
+The [Packager
+RFC](https://github.com/chadhietala/rfcs/blob/packager/active/0002-packager.md)
+submitted by [Chad Hietala](https://github.com/chadhietala) is a little over 2
+years old. A lot of things have changed since then and it requires a revision.
+
+The current application build process merges and concatenates input broccoli
+trees. This behaviour is not well documented and is a tribal knowledge. While
+the simplicity of this approach is nice, it doesn't allow for extension. We can
+refactor our build process and provide more flexibility when desired.
+
+Most importantly, the approach described below helps us achieve:
+
++ defining and developing a common language around the subject
++ removing highly coupled code and streamline technical implementation (Ember
+  Engines and Fastboot)
++ unlock a whole different set of plugins we couldn't have before:
+  + ability to create custom bundles (i.e per-engine and per-route bundles)
+  + take advantage of [HTTP2
+    multiplexing](https://http2.github.io/faq/#why-is-http2-multiplexed) and
+    [cache
+    pushing](https://www.mnot.net/blog/2014/01/30/http2_expectations#4-cache-pushing)
+  + optimising plugins (JavaScript and CSS tree-shaking)
+
+# Scope
+
++ New public API for customising build process and giving more granular control over
+the final build output
+
+# Terminology
+
++ **Packaging** - The process of designing, evaluating, and producing final build assets.
+
+# Detailed design
+
+The detailed design is separated in various sections so that it is easier for a
+reader to understand.
+
+## Packaging
+
+It gives you granular control over the final build output. It could be used in many
+different ways (we are going to go over use cases below). Note, it isn't meant to be
+used for "postprocess" transformations; "postprocess" is called after packaging is
+finished.
+
+Currently, Ember.js application and all of its depedencies get assembled under one
+directory with the following structure:
+
+```ruby
+bundler:js:input/
+├── addon-tree-output/
+├── the-app-name-folder/
+├── node_modules/
+└── vendor/
+```
+
+where:
+
++ `addon-tree-output` is a folder that contains dependencies from Ember add-ons
++ `the-app-name-folder` is a folder that contains Ember application code
++ `node_modules` is a folder that contains node dependencies
++ `vendor` is a folder that contains other dependencies.
```

Comment:
Have we thought about including the test tree here?

We may want to produce separate bundles for lazy loading tests between an engine and the host